File: hs/CirclesRandomSize.hs
{-# LANGUAGE ApplicativeDo     #-}
{-# LANGUAGE OverloadedStrings #-}
module CirclesRandomSize (benchmark, summary) where

import           Control.Monad  hiding (sequence_)
import           Data.Foldable  (for_)
import           Data.List      (zip4)
import           Data.Text      hiding (count)
import           Graphics.Blank
import           System.Random
import           Utils

benchmark :: CanvasBenchmark
benchmark ctx = do
    xs <- replicateM numCircles $ randomXCoord ctx
    ys <- replicateM numCircles $ randomYCoord ctx
    rs <- replicateM numCircles $ randomRIO (radiusMin, radiusMax)
    send' ctx $ for_ (zip4 xs ys rs (cycle ["red","blue","green"]))
                     (\ x y r col -> showBall (x, y) r col)


summary :: String
summary = "CirclesRandomSize"

numCircles :: Int
numCircles = 1000 * 1

radiusMin, radiusMax :: Double
radiusMin  = 1
radiusMax  = 50

showBall :: Point -> Double -> Text -> Canvas ()
showBall (x, y) r col = do
    beginPath();
    fillStyle(col);
    arc(x, y, r, 0, pi*2, False);
    closePath();
    fill();

File: js/CirclesRandomSize.js
function CirclesRandomSize(num, ctx) {
    var x;
    var y;
    var col;
    var RadiusMin = 1;
    var RadiusMax = 50;
    for (i = 0; i < num * 10; i++) {
        x = Math.floor(Math.random() * canvas.width);
        y = Math.floor(Math.random() * canvas.height);
        r = Math.floor(Math.random() * (RadiusMax - RadiusMin + 1) + RadiusMin);
        switch (i % 3) {
            case 0:
                col = "red";
                break;
            case 1:
                col = "blue";
                break;
            case 2:
                col = "green";
                break;
        }

        ctx.beginPath();
        ctx.fillStyle = col;
        ctx.arc(x, y, r, 0, 2 * Math.PI, false);
        ctx.closePath();
        ctx.fill();
    }
}

